### PR TITLE
[8.x] Add generics to Illuminate\Support\Collection

### PR DESCRIPTION
**Pull request by FrontEndCoffee**
Many users of Laravel use static analysis these days, and one of the most requested missing features of PHP are generics.

I thought adding them to `Collection` would be a great place to start with adding this functionality.

I don't know if this is something for Laravel 8.x or later, but I would love for these kinds of changes to be added in the framework. Especially with the new PHPStorm update coming up that adds better support (https://blog.jetbrains.com/phpstorm/2021/07/phpstorm-2021-2-beta/)

Let me know what you think about this, and if this is something that is in line with the vision of the Laravel framework. I would absolutely love it :D 

**Edit**: The `@template` syntax is supported by PHPStan as well as Psalm. It is also supported in the new version of PHPstorm and seems to be the default until this is implemented in the language.

### CODE REVIEW

File: src/Illuminate/Collections/Collection.php [hunk 1]
```diff
@@ -402,7 +405,7 @@ public function forget($keys)
      *
      * @param  mixed  $key
      * @param  mixed  $default
-     * @return mixed
+     * @return T
```
Comment (GrahamCampbell) [inline src/Illuminate/Collections/Collection.php:408]: this is wrong because it can also return the type of default

Comment (GrahamCampbell) [reply src/Illuminate/Collections/Collection.php:408]: its also wrong to just make default T... a common thing people do is default to a different type... like null :)

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
I don't think this belongs in the framework. Instead, should be shipped as 3rd party type defs.

File: src/Illuminate/Collections/Collection.php (hunk 1 above)
Comment (FrontEndCoffee) [reply src/Illuminate/Collections/Collection.php:408]: Hold on, I'll propose an implementation for this.

Comment (FrontEndCoffee) [reply src/Illuminate/Collections/Collection.php]: We can do something like https://github.com/laravel/framework/pull/38042/commits/0d324cede268abeb89d4c1b49b312ad798f9fe9a
, would that be acceptable?

**Comment by FrontEndCoffee** [comment]
> I don't think this belongs in the framework. Instead, should be shipped as 3rd party type defs.

I absolutely believe that this should be in the framework. It won't get in the way of people who don't use types, but for those that do it can be a lifesaver. I think this fits the vision of laravel pretty well.

**Comment by taylorotwell** [comment]
So what happens if the collection contains mixed types... for example a collection of strings and integers, or integers and floats, etc.?

**Comment by GrahamCampbell** [comment]
> for example a collection of strings and integers, or integers and floats, etc.?

`T` can be bound to `int|float|string` or even `mixed`.

**Comment by FrontEndCoffee** [comment]
Indeed, and if this is implemented throughout the framework, eventually HasMany relations can be typed as collections containing the model.

**Comment by FrontEndCoffee** [comment]
@taylorotwell https://psalm.dev/docs/annotating_code/templated_annotations/ this page describes pretty well how generics can already be implemented through annotations. This syntax is supported by PHPStan, Plasm and now by PHPStorm (and probably others).

File: src/Illuminate/Collections/Collection.php [hunk 2]
```diff
@@ -10,21 +10,24 @@
 use Illuminate\Support\Traits\Macroable;
 use stdClass;
 
+/**
+ * @template T
```
Comment (canvural) [inline src/Illuminate/Collections/Collection.php:14]: There should be two templates here. `TKey` and `TValue` maybe. So `offset*` methods can also use the key type.

Also in PHPStan (not sure about Psalm) `ArrayAccess` is also generic. So there needs to be `@implements ArrayAccess<TKey, TValue>` here.

Comment (FrontEndCoffee) [reply src/Illuminate/Collections/Collection.php:14]: Yes this crossed my mind but I wanted to keep the scope small.

However, I think it's way more usable if I add this. Great suggestion!

Comment (FrontEndCoffee) [reply src/Illuminate/Collections/Collection.php]: @canvural I have made the changes :D

File: src/Illuminate/Collections/Collection.php [hunk 3]
```diff
@@ -11,7 +11,8 @@
 use stdClass;
 
 /**
- * @template T
+ * @template TKey
```
Comment (GrahamCampbell) [inline src/Illuminate/Collections/Collection.php:14]: K and V are more conventional, here

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Collections/Collection.php (hunk 3 above)
Comment (FrontEndCoffee) [reply src/Illuminate/Collections/Collection.php]: @GrahamCampbell I've refactored it. 👍

**Comment by taylorotwell** [comment]
I'll probably need someone more entrenched in these concepts to properly review this for me.

**Comment by taylorotwell** [comment]
There are already community maintained implementations of this that have been reviewed and vetted more... I'm fine with adding support for this type of thing for sure, but maybe we can leverage the work that has already been done? 

https://github.com/nunomaduro/larastan/blob/master/stubs/Collection.stub

**Comment by FrontEndCoffee** [comment]
> There are already community maintained implementations of this that have been reviewed and vetted more... I'm fine with adding support for this type of thing for sure, but maybe we can leverage the work that has already been done?
> 
> https://github.com/nunomaduro/larastan/blob/master/stubs/Collection.stub

Good point, I'll take a look at it and update it if necessary. I would also like to implement it in more places, I could of course reuse the work already done in the larastan repo.

**Comment by taylorotwell** [comment]
Yeah those have been working solid for the Larastan community for a while.

**Comment by FrontEndCoffee** [comment]
@taylorotwell @GrahamCampbell I've taken the typehints used in the Larastan stubs and added these to the Collection and Enumerable classes. 

@GrahamCampbell said he prefers to use `K` and `V` as template names, however by looking at other projects, it seems that `TKey` and `TValue` is the common standard so I've implemented it like that.

**Comment by FrontEndCoffee** [comment]
I think it would be good if we checked type violations using PHPStan or Psalm in the framework, but I think that's would broaden the scope of this PR too much.

**Comment by taylorotwell** [comment]
Thanks for those updates... @nunomaduro what do you think about this?

**Comment by taylorotwell** [comment]
Are you able to get this working on the latest PHPStorm beta? I'm on the latest beta but don't seem to get any type-hint information when using generics:

CleanShot 2021-07-20 at 11 23 36@2x

**Comment by canvural** [comment]
@taylorotwell I think that's because PHPStorm does not understand `callable(TValue, TKey)` syntax [yet](https://youtrack.jetbrains.com/issue/WI-61500). But for example `$users->first()->` should autocomplete.

**Comment by nunomaduro** [comment]
After discussing a little bit offline with @canvural (co-author of [Larastan](https://github.com/nunomaduro/larastan)) and @taylorotwell, my opinion is the following:

The introduction to **serious type definitions** on Laravel's core should have a good and planned start. By migrating the solid stubs - more than 4,000,000 downloads - from Larastan at once in a one pull request that is followed by a test suite that proves that the type definitions are correct using static analysis - similar to what we have at Larastan.

We are both - @canvural and I - are available to start this task soon, and leave a really good foundation for future pull requests that offer type definitions to Laravel's core.

By good foundation, we mean to have  "standards" in place such as how to generics, how to use native types such as `array<string, string>` instead of `array` or `callable(): bool` instead of just `callable`. In addition, have a good test suite, that reuses the existing Laravel Test Suite, to validate if the types on the core even make sense.

**Comment by FrontEndCoffee** [comment]
@nunomaduro I think it's great that you and @canvural can get started on this soon! I also wholeheartedly agree that there should be a strong foundation including some way to validate the types. 

What do you think is best for this PR? Maybe closing it for now is a good idea. When this foundation is added, I'd love to help implementing types throughout the framework.

On the topic of the test suite, I think it would be a good idea to add PHPStan to the framework. As far as I have seen it, not everything from Larastan can be reused, as it expects an actual application to exist and be bootstrapped for some checks. Do you have any ideas?

**Comment by taylorotwell** [comment]
I guess I will close this for now if the Larastan stuff can be moved over soon.

**Pull request closed without merging**